### PR TITLE
Add service detail configuration tab

### PR DESCRIPTION
**Pull request by orlandohohmeier**
---
⚠️  _WIP: Tests missing for many of the components and I'm unhappy with the sub list rendering_

---

Introduce a service detail configuration tab. 

**AC**
- [x] Service detail configuration tab
- [x] Handle service versions data in DCOSStore 
- [x] Tests for all components
- [x] Proper config (sub) list styling

/cc @leemunroe

### CODE REVIEW

**Comment by mesosphere-web** [comment]
Reviewer: @mlunoe

File: src/js/components/ServiceDetailConfigurationTab.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+import Service from '../structs/Service';
+
+class ServiceConfigurationTab extends React.Component {
+
+  render() {
+    let {service} = this.props;
+    let {currentVersionID} = service.getVersionInfo();
+    return (
+      <ConfigurationView
+        headline={`Current Configuration (${currentVersionID})`}
```
Comment (pierlo-upitup) [inline src/js/components/ServiceDetailConfigurationTab.js]: `Current Version` seems more appropriate.

File: src/js/components/ConfigurationView.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+import DCOSStore from '../stores/DCOSStore'
+import DescriptionList from './DescriptionList';
+import Service from '../structs/Service';
+
+class ConfigurationView extends mixin(StoreMixin) {
+  constructor() {
+    super();
+
+    this.store_listeners = [{name: 'dcos', events: ['change']}];
```
Comment (pierlo-upitup) [inline src/js/components/ConfigurationView.js]: Isn't that a very generic event to listen to? Sounds like we need to check if the version actually changed in our `shouldComponentUpdate`?

File: src/js/components/ConfigurationView.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+      'Backoff Factor': configuration.backoffFactor,
+      'Backoff': configuration.backoffSeconds,
+      'Max Launch Delay': configuration.maxLaunchDelaySeconds,
+      'URIs': configuration.uris,
+      'User': configuration.user,
+      'Args': configuration.args,
+      'Version': configuration.version
+    };
+
+    return (<DescriptionList
```
Comment (pierlo-upitup) [inline src/js/components/ConfigurationView.js]:
```
return (
  <DescriptionList
```
?

File: src/js/components/ConfigurationView.js [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    super();
+
+    this.store_listeners = [{name: 'dcos', events: ['change']}];
+  }
+
+  render() {
+    let {headline, service, versionID} = this.props;
+    let configuration = service.getVersions().get(versionID);
+
+    if (!configuration) {
```
Comment (pierlo-upitup) [inline src/js/components/ConfigurationView.js]: `configuration == null`

File: src/js/components/ConfigurationView.js [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+      DCOSStore.fetchServiceVersion(service.getId(), versionID);
```
Comment (pierlo-upitup) [inline src/js/components/ConfigurationView.js]: We should perform this dance in `componentDidMount` and update the internal state with whatever Service Version the DCOSStore emits when ready by listening to the appropriate events on the Store. 
The render function should be "pure" and just concern itself to render whatever is in the state, without side-effects.

File: src/js/components/ConfigurationView.js [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+      'URIs': configuration.uris,
+      'User': configuration.user,
+      'Args': configuration.args,
+      'Version': configuration.version
+    };
+
+    return (<DescriptionList
+      hash={headerValueMapping}
+      headline={headline} />);
+  }
```
Comment (pierlo-upitup) [inline src/js/components/ConfigurationView.js]: Please refactor the `render` function to just check if `this.state.version` is set and either return the loading screen or the DescriptionList.

File: src/js/components/ServiceDetailConfigurationTab.js [hunk 7]
```diff
@@ -0,0 +1,25 @@
+import React from 'react';
+
+import ConfigurationView from './ConfigurationView';
+import Service from '../structs/Service';
+
+class ServiceConfigurationTab extends React.Component {
+
+  render() {
+    let {service} = this.props;
+    let {currentVersionID} = service.getVersionInfo();
```
Comment (pierlo-upitup) [inline src/js/components/ServiceDetailConfigurationTab.js]: Do you envision a case where we might want to pass the ConfigurationView a version that is not the current one? If not, I'd suggest to just pass it the `service`.

File: src/js/components/ConfigurationView.js [hunk 8]
```diff
@@ -0,0 +1,76 @@
+import React from 'react';
+import mixin from 'reactjs-mixin';
+import {StoreMixin} from 'mesosphere-shared-reactjs';
+
+import DCOSStore from '../stores/DCOSStore'
+import DescriptionList from './DescriptionList';
+import Service from '../structs/Service';
+
+class ConfigurationView extends mixin(StoreMixin) {
```
Comment (pierlo-upitup) [inline src/js/components/ConfigurationView.js]: I'd call this `ServiceConfigurationView`.

File: src/js/components/ConfigurationView.js (hunk 2 above)
Comment (pierlo-upitup) [inline src/js/components/ConfigurationView.js]: nvm – the `DCOS_CHANGE` event is actually emitted for the right things. I only have a problem with its name, but that's out of this PR.

File: src/js/components/DescriptionList.js [hunk 9]
```diff
@@ -23,11 +23,14 @@ class DescriptionList extends React.Component {
 
       // Check whether we are trying to render an object that is not a
       // React component
-      if (typeof value === 'object' && !Array.isArray(value) &&
-        value !== null && !React.isValidElement(value)) {
+      if (typeof value === 'object' && value !== null &&
+        !React.isValidElement(value)) {
 
+        // TODO (orlandohohmeier): indicate (e.g. indent) sub list for better ux
```
Comment (pierlo-upitup) [inline src/js/components/DescriptionList.js]: I'd suggest to json.stringify objects into a `<pre>` like in Marathon UI. No need for sublists.

File: src/js/components/ConfigurationView.js [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+      'Dependencies': configuration.dependencies,
+      'Labels': configuration.labels,
+      'Resource Roles': configuration.acceptedResourceRoles,
+      'Container': configuration.container,
+      'CPUs': configuration.cpus,
+      'Environment': configuration.env,
+      'Executor': 'todo',
+      'Health Checks': configuration.healthChecks,
+      'Instances': configuration.instances,
+      'IP Address': configuration.ipAddress,
```
Comment (pierlo-upitup) [inline src/js/components/ConfigurationView.js]: So all optional values will have a headline but no data?

File: src/js/components/ConfigurationView.js (hunk 8 above)
Comment (pierlo-upitup) [inline src/js/components/ConfigurationView.js]: Or even `ServiceDetailConfigurationView`

File: src/js/components/ServiceDetailConfigurationTab.js [hunk 11]
```diff
@@ -0,0 +1,25 @@
+import React from 'react';
+
+import ConfigurationView from './ConfigurationView';
+import Service from '../structs/Service';
+
+class ServiceConfigurationTab extends React.Component {
```
Comment (pierlo-upitup) [inline src/js/components/ServiceDetailConfigurationTab.js]: does not match filename

File: src/js/components/ServiceDetailConfigurationTab.js [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+
+  render() {
+    let {service} = this.props;
+    let {currentVersionID} = service.getVersionInfo();
+    return (
```
Comment (pierlo-upitup) [inline src/js/components/ServiceDetailConfigurationTab.js]: Add a newline before return

**Comment by leemunroe** [comment]
Nice one @orlandohohmeier. Just confirming you only want feedback on what is in the configuration tab for this PR (i.e. ignoring that the action buttons are missing and "Running" vs "Healthy")?

This is [this view](https://invis.io/3Q6PARCUD#/144153366_Services_V7_-Service_Detail-_Cassandra-_Configuration-).

* The "compare versions" feature is missing - we've talked about that being a nice to have and assuming it's a different effort, just noting it here for transparency
* If we're not going to compare versions, we'll still need a way to view historical versions and rollback to - can you confirm that is on the roadmap?
* Some of the formatting and visual hierarchy needs some style tweaking - maybe we can work on that together in SF e.g. it's not quite clear what is a heading at the minute
* What do we want to do about fields that have no value? I know this is something we've chatted about in the past, whether to leave empty, N/A, "Unspecified" or not show that row /cc @philipnrmn @pierlo-upitup 
* We could maybe cut down some spacing so the key:value pairs are closer together - chatted with @aldipower about this too, where we could use `column-2` instead of 3 (Sidenote: the styles and format we use here should be the same as what we use on the task detail view)
* For any JSON blobs, would love to see similar `<pre>` formatting carried through from what we have in Marathon today. Not in the mockup but I was definitely a fan of that.

image

File: src/js/components/ConfigurationView.js (hunk 5 above)
Comment (mlunoe) [inline src/js/components/ConfigurationView.js]: Make that `componentWillMount`

File: src/js/components/DescriptionList.js [hunk 13]
```diff
@@ -23,11 +23,14 @@ class DescriptionList extends React.Component {
 
       // Check whether we are trying to render an object that is not a
       // React component
-      if (typeof value === 'object' && !Array.isArray(value) &&
-        value !== null && !React.isValidElement(value)) {
+      if (typeof value === 'object' && value !== null &&
```
Comment (mlunoe) [inline src/js/components/DescriptionList.js]: Why the change to accept arrays now? I'm sure there was a reason for it being there before.

Comment (mlunoe) [inline src/js/components/DescriptionList.js]: This is used in different places, so let's check it still works as expected there

File: src/js/components/ConfigurationView.js (hunk 8 above)
Comment (orlandohohmeier) [inline src/js/components/ConfigurationView.js]: I'd stick with the current naming as to be consistent with the `ServiceDetailTaskTab` and `TaskView` naming.

File: src/js/components/DescriptionList.js (hunk 9 above)
Comment (orlandohohmeier) [inline src/js/components/DescriptionList.js]: I opted for the sublists as it seemed easier than finding a way to properly render JSON blocks within a DescriptionList.

File: src/js/components/ConfigurationView.js (hunk 10 above)
Comment (orlandohohmeier) [inline src/js/components/ConfigurationView.js]: Yeah, just as in the current Marathon UI version...

File: src/js/components/DescriptionList.js (hunk 13 above)
Comment (orlandohohmeier) [inline src/js/components/DescriptionList.js]: I wasn't able to so spot any issues and all the tests passed...

File: src/js/components/ConfigurationView.js (hunk 5 above)
Comment (orlandohohmeier) [inline src/js/components/ConfigurationView.js]: Oh, yeah! This is a WIP implementation, as I wasn't sure how `componentWillMount` will work in combination with the `StoreMixin`. I will definitely refactor this!

File: src/js/components/ServiceDetailConfigurationTab.js (hunk 7 above)
Comment (orlandohohmeier) [inline src/js/components/ServiceDetailConfigurationTab.js]: Yeah, there we will an extend version of the configuration tab to display not only the current but also older versions.

File: src/js/components/ServiceDetailConfigurationTab.js (hunk 1 above)
Comment (orlandohohmeier) [inline src/js/components/ServiceDetailConfigurationTab.js]: "Current Configuration" is what the design says, but I'll update it.

File: src/js/components/ConfigurationView.js [hunk 14]
```diff
@@ -0,0 +1,76 @@
+import React from 'react';
+import mixin from 'reactjs-mixin';
+import {StoreMixin} from 'mesosphere-shared-reactjs';
```
Comment (kennyt) [inline src/js/components/ConfigurationView.js]: :abc: these imports :)

File: src/js/components/DescriptionList.js (hunk 13 above)
Comment (kennyt) [inline src/js/components/DescriptionList.js]: well we should definitely keep the array check. this is because `value` cannot be an array.

Comment (orlandohohmeier) [inline src/js/components/DescriptionList.js]: It's transformed into an object (`Object.assign({}, value)`) so that we will get sublists for arrays as well. But @leemunroe is currently exploring different ways to render the configuration details and I might drop this change...

Comment (kennyt) [inline src/js/components/DescriptionList.js]: so then the keys would be the `index` and the value would be the values in the array? do we want that?

Comment (orlandohohmeier) [inline src/js/components/DescriptionList.js]: We could join the index and the parent key/heading, but I will not iterate on this until we found a visually appealing way as the current styling neither works for "normal" object sublist nor array based sublist. So this change might get dropped.

File: src/js/components/DescriptionList.js (hunk 9 above)
Comment (pierlo-upitup) [inline src/js/components/DescriptionList.js]: And what about https://github.com/dcos/dcos-ui/pull/155#issuecomment-218921707 ?

**Comment by pierlo-upitup** [comment]
Let me know once the tests are ready I'll be happy to give this another pass.

File: src/js/components/ConfigurationView.js [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+
+function fetchVersion(service, versionID) {
+  if (service.getVersions().get(versionID) == null) {
+    DCOSStore.fetchServiceVersion(service.getId(), versionID);
+  }
+}
+
+class ConfigurationView extends mixin(StoreMixin) {
+  constructor() {
+    super();
```
Comment (jfurrow) [inline src/js/components/ConfigurationView.js]: We should pass `...arguments` to `super`

**Comment by orlandohohmeier** [comment]
I'll split this into two separate PRs as it grew into a beast...

**Comment by orlandohohmeier** [comment]
I've splitted the PR into smaller chunks. Please review #220, #221, and #222 instead.
